### PR TITLE
Missing hero's flag animation in a world map

### LINKED ISSUE #485: Missing hero's flag animation in a world map
We should add animation to hero's flag in a world map. Now it is static and looks like a simple triangle over a hero's head.
World flag

### PR DESCRIPTION
**Pull request by shprotru**
Fix #485

### CODE REVIEW

**Comment by shprotru** [comment]
Hi, @ihhub, I have changed additional [strange place](https://github.com/ihhub/fheroes2/pull/693/files#diff-1401bd1db7694136356108aa689d59ceR988) in existing code, we have identifiers HEROFL00 - HEROFL06, but reference only to HEROFL00 - HEROFL05, I have change it to use all range, till 06, logically it's correct and seems nothing braked.

File: src/fheroes2/agg/icn.cpp [hunk 1]
```diff
@@ -985,7 +985,8 @@ u32 ICN::AnimationFrame( int icn, u32 start, u32 ticket, bool quantity )
     case HEROFL03:
     case HEROFL04:
     case HEROFL05:
-        return ticket % 5;
+    case HEROFL06:
+        return ticket % 6;
```
Comment (ihhub) [inline src/fheroes2/agg/icn.cpp:989]: Hi @shprotru , this case is incorrect. `case HEROFL06:` is a correct change but **ticket % 6;** is not as each of **HEROFL00.. HEROFL06** contains only 5 frames. Please correct this place.

Comment (shprotru) [reply src/fheroes2/agg/icn.cpp]: done

**Comment by ihhub** [comment]
Hi @shprotru , I was trying to fix the same place :) The problem as seen on the screenshot that the flag is moving while it should be on fixed position:
image

File: src/fheroes2/agg/icn.cpp [hunk 2]
```diff
@@ -985,6 +985,7 @@ u32 ICN::AnimationFrame( int icn, u32 start, u32 ticket, bool quantity )
     case HEROFL03:
     case HEROFL04:
     case HEROFL05:
+    case HEROFL06:
```
Comment (ihhub) [inline src/fheroes2/agg/icn.cpp:988]: Hi @shprotru , could you please make a separate pull request just for this change as current changes do no work properly yet.

This change is needed for flag animation of neutral captain in a castle/town. Thanks for fixing this place!

**Comment by shprotru** [comment]
@ihhub attempt to fix [this](https://github.com/ihhub/fheroes2/pull/693#issuecomment-636772190) dirty way, to clean fix I think, is need to know right animation frames. Check how it's look now, better and incorrect or correct and acceptable?

File: src/fheroes2/agg/icn.cpp (hunk 2 above)
Comment (shprotru) [reply src/fheroes2/agg/icn.cpp]: [done](https://github.com/ihhub/fheroes2/pull/695)

**Comment by shprotru** [comment]
Hi @ihhub, seems it's done now, I made calculation tables for offset on Y axis. Check, please.

**Comment by shprotru** [comment]
Hi @ihhub, something wrong with building on `travis`, on my machine project build successfully. Seems here need your C++ magic trick.

File: src/fheroes2/heroes/heroes.cpp [hunk 3]
```diff
@@ -105,6 +105,11 @@ Heroes::Heroes()
     , sprite_index( 18 )
     , patrol_square( 0 )
     , _alphaValue( 255 )
+    , flagVOffsetTableBottom( std::vector<short> {0, 1, 2, 0, 1, 2, 3, 0, 1} )
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes.cpp:108]: Hi @shprotru , we do not support C++11 so we cannot use initialization lists on containers. This is why our compilation fails.

File: src/fheroes2/heroes/heroes.h [hunk 4]
```diff
@@ -338,6 +338,11 @@ class Heroes : public HeroBase, public ColorBase
     int patrol_square;
 
     std::list<IndexObject> visit_object;
+    std::vector<short> flagVOffsetTableBottom;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes.h:341]: Please use `int16_t` type instead.

Comment (ihhub) [reply src/fheroes2/heroes/heroes.h:341]: Why do we have an offset only for Y axis? I observed that we have a very noticeable offset for X axis.

File: src/fheroes2/heroes/heroes.cpp (hunk 3 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes.cpp]: It was working on C++03 for gcc, that's why I use this in hope that it could work. Returning back to idea of statically initialized array, this is common practice in C++, no need to make `push_back(value)` snake in this case.

Additionally in source `heroes_move.cpp` would be great to see this functions: `ReflectSprite`, 
`SpriteHero`, `SpriteFlag`, `SpriteShad`, `SpriteFroth` as class methods, what do you think on this?

File: src/fheroes2/heroes/heroes.h (hunk 4 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes.h]: done with int16_t.
will make attempt to add `X` axis table

**Comment by shprotru** [comment]
Hi @ihhub, check, please.

File: src/fheroes2/heroes/heroes.h [hunk 5]
```diff
@@ -27,13 +27,16 @@
...
 
+#include "agg.h"
 #include "army.h"
 #include "gamedefs.h"
 #include "heroes_base.h"
 #include "pairs.h"
 #include "route.h"
 #include "visit.h"
 
+#define amountOfFlagFrames 9
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes.h:38]: Please make this value as a private unnamed enumeration within Heroes class:
```
enum
{
    FLAG_FRAME_COUNT = 9
};
```

File: src/fheroes2/heroes/heroes_move.cpp [hunk 6]
```diff
@@ -352,8 +352,65 @@ void Heroes::Redraw( Surface & dst, s32 dx, s32 dy, bool with_shadow ) const
...
+    static const int16_t flagYOffsetTableBottom[amountOfFlagFrames] = {0, 1, 2, 0, 1, 2, 3, 0, 1};
+    static const int16_t flagYOffsetTableTop[amountOfFlagFrames] = {0, -2, -3, -2, 0, -1, -3, -2, -1};
+    static const int16_t flagYOffsetTableBottomAndSideways[amountOfFlagFrames] = {0, 0, 1, 2, 2, 3, 2, 1, 0};
+    static const int16_t flagYOffsetTableTopAndSideways[amountOfFlagFrames] = {0, 0, 0, 1, 1, 0, 0, 0, 0};
+    static const int16_t flagYOffsetTableSideways[amountOfFlagFrames] = {0, 0, 0, 0, 1, 1, 0, 0, 0};
+
+    static const int16_t flagXOffsetTableBottom[amountOfFlagFrames] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
+    static const int16_t flagXOffsetTableTop[amountOfFlagFrames] = {0, 0, 0, 0, 0, 0, 0, 0, 0};
+    static const int16_t flagXOffsetTableBottomAndRight[amountOfFlagFrames] = {0, 0, -1, -3, -1, 0, 0, 0, -2};
+    static const int16_t flagXOffsetTableBottomAndLeft[amountOfFlagFrames] = {0, -1, 0, 2, 0, -1, -1, -1, 1};
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:365]: Left side must have opposite values of right side as we just reflect a frame. It is true for top but somehow its is different here for bottom.

File: src/fheroes2/heroes/heroes_move.cpp [hunk 7]
```diff
@@ -352,8 +352,65 @@ void Heroes::Redraw( Surface & dst, s32 dx, s32 dy, bool with_shadow ) const
         sprite4.SetAlphaMod( _alphaValue );
     }
 
+    int16_t *offsettableX, *offsettableY;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:355]: Let's move whole this code below to `SpriteFlag` function and assign needed position of a frame by *SetPos()* method.

File: src/fheroes2/heroes/heroes.h (hunk 5 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes.h:38]: done. I made a decision make this 4 functions as class methods, because `SpriteFlag` used this constant.

File: src/fheroes2/heroes/heroes_move.cpp (hunk 6 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes_move.cpp]: something to do additional with that?

File: src/fheroes2/heroes/heroes_move.cpp (hunk 7 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes_move.cpp]: done

File: src/fheroes2/heroes/heroes.cpp (hunk 3 above)
Comment (ihhub) [reply src/fheroes2/heroes/heroes.cpp]: It is not a good idea as Heroes class should not contain drawing related stuff.

Comment (shprotru) [reply src/fheroes2/heroes/heroes.cpp]: FLAG_FRAME_COUNT is class-related definition, how to make it accessible in SpriteFlag? Heroes class already contain `Redraw` method, so we need separate this method and `Heroes` class, could you help in that?

Comment (ihhub) [reply src/fheroes2/heroes/heroes.cpp]: Hi @shprotru , we are almost there!

Let's do these steps:
1. Revert all changes made in `heroes.h` file
2. Revert changes made for these 4 functions
3. Modify `SpriteFlag` function like this:
```
Sprite SpriteFlag( const Heroes & hero, int index, int frameId, bool reflect, bool rotate )
{
    int icn_flag = ICN::UNKNOWN;
    int index_sprite = 0;

    switch ( hero.GetColor() ) {
    case Color::BLUE:
        icn_flag = hero.isShipMaster() ? ICN::B_BFLG32 : ICN::B_FLAG32;
        break;
    case Color::GREEN:
        icn_flag = hero.isShipMaster() ? ICN::G_BFLG32 : ICN::G_FLAG32;
        break;
    case Color::RED:
        icn_flag = hero.isShipMaster() ? ICN::R_BFLG32 : ICN::R_FLAG32;
        break;
    case Color::YELLOW:
        icn_flag = hero.isShipMaster() ? ICN::Y_BFLG32 : ICN::Y_FLAG32;
        break;
    case Color::ORANGE:
        icn_flag = hero.isShipMaster() ? ICN::O_BFLG32 : ICN::O_FLAG32;
        break;
    case Color::PURPLE:
        icn_flag = hero.isShipMaster() ? ICN::P_BFLG32 : ICN::P_FLAG32;
        break;

    default:
        DEBUG( DBG_GAME, DBG_WARN, "unknown color" );
        break;
    }

    if ( rotate )
        index_sprite = 45;
    else
        switch ( hero.GetDirection() ) {
        case Direction::TOP:
            index_sprite = 0;
            break;
        case Direction::TOP_RIGHT:
            index_sprite = 9;
            break;
        case Direction::RIGHT:
            index_sprite = 18;
            break;
        case Direction::BOTTOM_RIGHT:
            index_sprite = 27;
            break;
        case Direction::BOTTOM:
            index_sprite = 36;
            break;
        case Direction::BOTTOM_LEFT:
            index_sprite = 27;
            break;
        case Direction::LEFT:
            index_sprite = 18;
            break;
        case Direction::TOP_LEFT:
            index_sprite = 9;
            break;

        default:
            DEBUG( DBG_GAME, DBG_WARN, "unknown direction" );
            break;
        }

    static const Point offsetTop[heroFlagFrameCount] = {Point( 0, 0), ...};
    static const Point offsetTopRight[heroFlagFrameCount] = {Point( 0, 0), ...};
    static const Point offsetRight[heroFlagFrameCount] = {Point( 0, 0), ...};
    static const Point offsetBottomRight[heroFlagFrameCount] = {Point( 0, 0), ...};
    static const Point offsetBottom[heroFlagFrameCount] = {Point( 0, 0), ...};

    // for Bottom-left, left and top-left we must use negative values from right side
    Point offset;
    switch ( GetDirection() ) {
    case Direction::TOP:
        offset = offsetTop[frameId];
        break;
    case Direction::BOTTOM:
       ...
    case Direction::LEFT:
       offset = -offsetRight[frameId];
    }

    Sprite flag = AGG::GetICN( icn_flag, index_sprite + ( index % 9 ), reflect );
    flag.SetPos( flag.Pos() + offset );
    return flag;
}
```
4. Declare:
```
namespace
{
    const int heroFlagFrameCount = 9;
}
```

As a result almost all changes will be only in `SpriteFlag` function.

Comment (shprotru) [reply src/fheroes2/heroes/heroes.cpp]: Hi, @ihhub got `error: no match for ‘operator-’ (operand type is ‘const Point’)
         offset = -offsetTop[frameId];`
Need I wrote overload operator for `-` ?

Comment (ihhub) [reply src/fheroes2/heroes/heroes.cpp]: You could overload this operator.

Comment (shprotru) [reply src/fheroes2/heroes/heroes.cpp]: Hi, @ihhub is values for top and bottom directions should be invertable?

Comment (ihhub) [reply src/fheroes2/heroes/heroes.cpp]: No :)

Comment (shprotru) [reply src/fheroes2/heroes/heroes.cpp]: @ihhub hope, i done. Can't get how it works without `-` arithmetic, but seems this is work. Check, please.

**Comment by ihhub** [comment]
Hi @shprotru , all seems good except the fact that these changes should be applied only for a flag of a hero on land, not for a boat of see... I will come up with a solution tomorrow as it is pretty late for me right now.

**Comment by shprotru** [comment]
@ihhub a little higher I saw the calls for `hero.isShipMaster()`, seems need to use it. Strange frame 5 when a boat in bottom and sideways position, it has `248` pixels height, could I ask you inspect this frame?

**Comment by ihhub** [comment]
> @ihhub a little higher I saw the calls for `hero.isShipMaster()`, seems need to use it. Strange frame 5 when a boat in bottom and sideways position, it has `248` pixels height, could I ask you inspect this frame? Is it correct?

I saw multiple frames like this. That's why it's highly not recommended to use w() and h() of a Sprite. However, x() and y() will compensate such image size.

**Comment by shprotru** [comment]
Hi, @ihhub! I noticed that in OG if hero_in_a_boat does not move's, flag must be static. And logic on animating flag on movement of hero_in_a_boat is already exists.

**Comment by ihhub** [comment]
> Hi, @ihhub! I noticed that in OG if hero_in_a_boat does not move's, flag must be static. And logic on animating flag on movement of hero_in_a_boat is already exists.

With current changes the flat on boat always moves. What I am proposing is to remove animation for this flag at all. When you move a boat you cannot see this animation properly.

**Comment by shprotru** [comment]
Hi @ihhub!
> With current changes the flat on boat always moves.

No, it does not, because of `68ad460 `:
```
if ( hero.isShipMaster() )
        frameId = 0;
```
Do you mean some kind of additional code change?

**Comment by ihhub** [comment]
Hi @shprotru , please pull latest changes from master branch. It is not easy to see if flag is on correct place. I noticed two problems:
- when I move a hero facing left to top or bottom left corner the flag becomes dis-attached for 1 or 2 frames
- during boat movement we have see 2 flags :)

**Comment by shprotru** [comment]
Hi @ihhub! How you visually get dis-attaching of flag on left diagonal directions? This is caused by horizontal offset? I broken my eyes at this moment, don't know how to debug, it looks good for me.
Understand the moment about the boat:
each frame of the boat contain it's own neutral flag, the hero's(colour) flag should be drawn on top of boat's frame, that's why we could see 2 flags per 1 frame(additionally they should be different colors), dissynchronization of animation.
Can't call for `SetShipMaster( true );` with debugging purposes in `Redraw` method because it's `const`.

**Comment by ihhub** [comment]
> Hi @ihhub! How you visually get dis-attaching of flag on left diagonal directions? This is caused by horizontal offset? I broken my eyes at this moment, don't know how to debug, it looks good for me.
> Understand the moment about the boat:
> each frame of the boat contain it's own neutral flag, the hero's(colour) flag should be drawn on top of boat's frame, that's why we could see 2 flags per 1 frame(additionally they should be different colors), dissynchronization of animation.
> Can't call for `SetShipMaster( true );` with debugging purposes in `Redraw` method because it's `const`.

Hmmm... now I don't see it :( Let's fix boat first.

**Comment by shprotru** [comment]
Hi, @ihhub! I've done with ship flag position tables and seems I've found a bug:
1. Load a game with visible map
2. End turn
3. Watch on AI controlled heroes, you could notice that when they are finished their movement the function `isEnableMove()` will always return true.
Looks like a bug, could you help on inspect it? This bug not related to code changes in this PR.

I notice 4 states:
- [x] hero static, flag dynamic
- [x] hero in movement, flag dynamic
- [x] boat static, flag static
- [x] boat in movement, flag dynamic

Seems everything done, except described bug.

**Comment by shprotru** [comment]
Hi @ihhub, could you help understand this small strange moments:

1. ~~[first](https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/ai/ai_hero_action.cpp#L1899), could you remove this `if` condition?~~
2. no tab [here](https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/gui/interface_events.cpp#L397) and strange formatting [here](https://github.com/ihhub/fheroes2/blob/master/src/fheroes2/gui/interface_events.cpp#L405)

found missing `SetMove` call, so everything should work fine, check, please.

**Comment by shprotru** [comment]
Hi, @ihhub! Could you check this PR, is it still have problems?

File: src/fheroes2/ai/ai_hero_action.cpp [hunk 8]
```diff
@@ -1947,6 +1947,8 @@ namespace AI
                 if ( !hero.isFreeman() && !hide_move )
                     DELAY( 200 );
             }
+
+            hero.SetMove( false );
```
Comment (ihhub) [inline src/fheroes2/ai/ai_hero_action.cpp:1951]: Why do we need this change?

File: src/fheroes2/heroes/heroes_move.cpp [hunk 9]
```diff
@@ -335,14 +389,24 @@ void Heroes::Redraw( Surface & dst, bool with_shadow ) const
 void Heroes::Redraw( Surface & dst, s32 dx, s32 dy, bool with_shadow ) const
 {
     const Point & mp = GetCenter();
+    int heroFrameID = Game::MapsAnimationFrame() % heroFrameCount;
```
Comment (ihhub) [inline src/fheroes2/heroes/heroes_move.cpp:392]: Why do we need such change? Hero's movement now became in 3-4 times slower while speed remains the same.

File: src/fheroes2/ai/ai_hero_action.cpp (hunk 8 above)
Comment (shprotru) [reply src/fheroes2/ai/ai_hero_action.cpp:1951]: Yes, we need, because the AI controlled hero, after finish his movement, does not reset flag `ENABLEMOVE`, which could affect on animation in redraw cycle if you check his state with `isEnableMove` call. This flag does not must indicate that hero is `in movement` state, after his turn is finished. What do you think on this?

File: src/fheroes2/heroes/heroes_move.cpp (hunk 9 above)
Comment (shprotru) [reply src/fheroes2/heroes/heroes_move.cpp:389]: Yeah, I got that, can't understand what is it and how to do with that. I've done it a tricky way: simply synchronize a hero animation frame with flag animation frame. Seems this moment done in OG this way! Now looks much better! Check again, please. Sorry for the frequent worries, I am often need that somebody additionally look that I am looked and share his thoughts with me.

Comment (ihhub) [reply src/fheroes2/heroes/heroes_move.cpp:389]: No problem at all! Nobody blames anyone ;)

I noticed that with latest changes the boat animation moves in 2 times slower. Is it the same as in the original game?

Comment (shprotru) [reply src/fheroes2/heroes/heroes_move.cpp]: Hi @ihhub in OG it's as you say, flag is fast animated, changed it as you ask, check again, please.

File: src/fheroes2/ai/ai_hero_action.cpp (hunk 8 above)
Comment (ihhub) [reply src/fheroes2/ai/ai_hero_action.cpp:1950]: `hero.Move();` will call this code internally. Maybe after all our changes (we modify the logic a bit) we do not need it. Could you please verify this?

File: src/fheroes2/heroes/heroes_move.cpp [hunk 10]
```diff
@@ -227,7 +232,54 @@ Sprite SpriteFlag( const Heroes & hero, int index, bool reflect, bool rotate )
             break;
         }
 
-    return AGG::GetICN( icn_flag, index_sprite + ( index % 9 ), reflect );
+    const int frameId = index % heroFrameCount;
+    Sprite flag = AGG::GetICN( icn_flag, index_sprite + frameId, reflect );
+    if ( !hero.isEnableMove() ) {
```
Comment (shprotru) [inline src/fheroes2/heroes/heroes_move.cpp:237]: How could I not have guessed! Good code style improvement here

File: src/fheroes2/heroes/heroes_move.cpp [hunk 11]
```diff
@@ -341,8 +393,13 @@ void Heroes::Redraw( Surface & dst, s32 dx, s32 dy, bool with_shadow ) const
 
     bool reflect = ReflectSprite( direction );
 
+    int flagFrameID = sprite_index;
+    if ( !isEnableMove() ) {
+        flagFrameID = isShipMaster() ? 0 : Game::MapsAnimationFrame();
```
Comment (shprotru) [inline src/fheroes2/heroes/heroes_move.cpp:398]: Nice change!

File: src/fheroes2/ai/ai_hero_action.cpp (hunk 8 above)
Comment (shprotru) [reply src/fheroes2/ai/ai_hero_action.cpp:1950]: We still need this change, because we use call for `hero.isEnableMove()`, to check hero movement state, this is depend on `ENABLEMOVE` flag. Seems everything looks good with this change, can't be sure about AI, but as I say, everything looks good. There is a 100% bug in existing code base around `ENABLEMOVE` flag and here we fix it.
When AI controlled hero finished his turn - we must reset `ENABLEMOVE` flag, because he finished his movement. Interesting to see `partially movement` condition of AI, if it's possible:

AI controlled hero move for 4 squares, after that he change army with other AI controlled hero from same player. Could he continue his movement? I don't know: is it possible and how to check this condition.

**Comment by shprotru** [comment]
Hi @ihhub! After your changes everything looks good for me you nicely improved my code style.

**Comment by ihhub** [comment]
Hi @shprotru , thank you very much for your non-stop contribution!
